web/CMakeLists.txt

```diff
@@ -1,78 +1,45 @@
 project(trweb)
 
-set(TRWEB_SRCS
-  src/about-dialog.js
-  src/action-manager.js
-  src/alert-dialog.js
-  src/context-menu.js
-  src/file-row.js
-  src/formatter.js
-  src/inspector.js
-  src/main.js
-  src/move-dialog.js
-  src/notifications.js
-  src/open-dialog.js
-  src/overflow-menu.js
-  src/prefs-dialog.js
-  src/prefs.js
-  src/remote.js
-  src/remove-dialog.js
-  src/rename-dialog.js
-  src/shortcuts-dialog.js
-  src/statistics-dialog.js
-  src/torrent.js
-  src/torrent-row.js
-  src/transmission.js
-  src/utils.js
-  style/transmission-app.scss
+file(GLOB_RECURSE TRWEB_ASSETS
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/assets/img/*.png
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/assets/img/*.svg
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/assets/img/favicon.ico
 )
 
-set(TRWEB_IMGS
-  style/images/analytics.svg
-  style/images/application-x-executable.png
-  style/images/audio-x-generic.png
-  style/images/blue-turtle.png
-  style/images/checkered-flag.svg
-  style/images/chevron-down.svg
-  style/images/chevron-up.svg
-  style/images/diagram-3-fill.svg
-  style/images/files.svg
-  style/images/folder.png
-  style/images/font-x-generic.png
-  style/images/gear-fill.svg
-  style/images/horizontal-rule.svg
-  style/images/image-x-generic.png
-  style/images/lock-fill.svg
-  style/images/logo.png
-  style/images/package-x-generic.png
-  style/images/pause-circle-active.svg
-  style/images/pause-circle-idle.svg
-  style/images/play-circle-active.svg
-  style/images/play-circle-idle.svg
-  style/images/router.svg
-  style/images/search.svg
-  style/images/team.svg
-  style/images/text-x-generic.png
-  style/images/three-dots-vertical.svg
-  style/images/toolbar-close.png
-  style/images/toolbar-folder.png
-  style/images/toolbar-info.png
-  style/images/toolbar-pause.png
-  style/images/toolbar-start.png
-  style/images/turtle.png
-  style/images/up-and-down-arrows.svg
-  style/images/video-x-generic.png
+file(GLOB_RECURSE TRWEB_TMP_BUILD
+  ${CMAKE_CURRENT_SOURCE_DIR}/public_html/assets/**/*
+  ${CMAKE_CURRENT_SOURCE_DIR}/public_html/js/*
+  ${CMAKE_CURRENT_SOURCE_DIR}/public_html/*
 )
 
-add_custom_target(
-  trweb ALL
-  COMMAND ${CMAKE_COMMAND} -E copy "${CMAKE_CURRENT_SOURCE_DIR}/package.json" "${CMAKE_CURRENT_BINARY_DIR}/package.json"
-  COMMAND ${CMAKE_COMMAND} -E copy "${CMAKE_CURRENT_SOURCE_DIR}/yarn.lock" "${CMAKE_CURRENT_BINARY_DIR}/yarn.lock"
-  COMMAND yarn install
-  COMMAND yarn webpack --config "${CMAKE_CURRENT_SOURCE_DIR}/webpack.config.js" --context "${CMAKE_CURRENT_SOURCE_DIR}"
-  BYPRODUCTS
-    public_html/transmission-app.js
-    public_html/transmission-app.js.LICENSE.txt
-  DEPENDS ${TRWEB_IMGS}
-  SOURCES ${TRWEB_SRCS}
+file(GLOB_RECURSE TRWEB_SRCS
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/*.js
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/index.html
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/assets/css/*.scss
 )
+
+find_program(NODE node)
+
+if (NODE)
```

Comment:
If `ENABLE_WEB` is turned on, any failure in finding the tools necessary for the build (as well as any other failures) should result in an error. Right now I cannot be sure that WEB component will in fact be build if I set `ENABLE_WEB` to `ON` since this check silently ignores the request.

If you want to only build WEB component if the necessary build tools were bound, I think it would be better to do the same as for the other major components, i.e. support (and default to) an `AUTO` option. `ON` should mean `ON`, `OFF` should mean `OFF`.

Reply:
I set it to `ON` under the rationale that these are files that are served up by the other apps, e.g. transmission-daemon, so if GTK / Qt / daemon are built, we ought to be as well. And since daemon is set to `ON`, that means web should be too.

Probably best to remove `if (NODE)` here as redundant with `ENABLE_WEB`

Reply:
To me web is just another client. It's true that other clients have a link to it, but it doesn't necessarily mean that it needs to be built when they are. Some people use other web (or non-web remote) clients so they may not be interested in installing the default one.

Reply:
That's true; and if so, they can opt out of the web client by changing the setting to 'off'.

If the _default_ is to build the daemon but only build the web client iff node is installed, then users are going to send us bug reports because some packager somewhere didn't know that they needed to have Node installed :)

Reply:
Well, `AUTO` also gives you an option to explicitly opt-in :) Maybe it's just me but I think for package maintainers being explicit either way its always a good idea (see e.g. [Gentoo package](https://gitweb.gentoo.org/repo/gentoo.git/tree/net-p2p/transmission/transmission-3.00-r1.ebuild#n87)), otherwise you don't know what you're building...

I was thinking more in terms of CI, where I don't want to control whether it's on or off except for builds I really care about, like e.g. Mac and Windows where it's included into the distributable packages. Building it on all the bots makes little sense since the result will unlikely be different (in a sense, Web is a single platform), but then explicitly disabling it on most of the platform is just more work that I'd rather not do :-P (yeah, I'm that lazy at times)

Daemon and utils are kind of different since they don't really require anything on top of what libtr already requires, so enabling them will never (?) break the build. My thinking behind introducing `AUTO` was so that the build doesn't fail in most cases if you don't have dependencies (e.g. GTK, Qt) because I hate installing additional stuff for something I may not even need; if you're a developer you'll notice things are missing after default configuration and you'll (hopefully) figure it out. I'm open to improving CMake messages issued, e.g. telling user that something that's `AUTO` isn't going to be built because something is missing (I don't think this is happening right now).

Looks like this may be a bigger discussion outside of this PR, so I'm fine with leaving the default of `ON` for now, but only if it fails the build when `node` is missing.